### PR TITLE
MEN-4882: Query the NeedsArtifactReboot parameter in rollback

### PR DESCRIPTION
**Pull request by oleorhagen**
Changelog: None
Signed-off-by: Ole Petter <ole.orhagen@northern.tech>

### CODE REVIEW

**Comment by oleorhagen** [comment]
@mender-test-bot start pipeline

File: app/state.go [hunk 1]
```diff
@@ -1673,6 +1673,47 @@ func (rs *updateRollbackState) Handle(ctx *StateContext, c Controller) (State, b
...
+
+	// Query the rollback parameter from the Update Module, in case it has
+	// not been called previously (MEN-4882)
+	for n, i := range c.GetInstallers() {
+		v, err := rs.Update().RebootRequested.Get(n)
+		if err != nil {
+			log.Errorf(
+				"Failed to query the database for the Update Module's"+
+					" ArtifactRollback parameter during updateRollbackState: %s"+
+					"Will query the update module anew.",
```
Comment (kacf) [inline app/state.go:1685]: No space/punctuation between these two lines.

File: app/state.go [hunk 2]
```diff
@@ -1673,6 +1673,47 @@ func (rs *updateRollbackState) Handle(ctx *StateContext, c Controller) (State, b
 			return rs.HandleError(ctx, c, NewFatalError(err))
 		}
 	}
+
+	// Query the rollback parameter from the Update Module, in case it has
+	// not been called previously (MEN-4882)
+	for n, i := range c.GetInstallers() {
+		v, err := rs.Update().RebootRequested.Get(n)
+		if err != nil {
+			log.Errorf(
```
Comment (kacf) [inline app/state.go:1682]: How about dialling this back to `Debugf`? After this fix it's not a bug/error anymore, so I don't think it's necessary to be loud about it.

File: app/state.go [hunk 3]
```diff
@@ -1673,6 +1673,47 @@ func (rs *updateRollbackState) Handle(ctx *StateContext, c Controller) (State, b
...
+				"Failed to query the database for the Update Module's"+
+					" ArtifactRollback parameter during updateRollbackState: %s"+
+					"Will query the update module anew.",
+				err.Error())
+		}
+		switch v {
+		// If the value is 'RebootRequested', or 'RebootTypeAutomatic',
+		// do not risk writing a 'RebootTypeNone', reboot in any case
```
Comment (kacf) [inline app/state.go:1690]: 👍

File: app/state_test.go [hunk 4]
```diff
@@ -5405,3 +5409,53 @@ func TestFetchRetryUpdateControl(t *testing.T) {
...
+
+	for name, testCase := range tests {
+		t.Logf("Running testcase:  %s", name)
+		ms := store.NewMemStore()
+		ctx := &StateContext{
+			Store: ms,
+		}
+		c := &stateTestController{}
+
+		c.FakeDevice.NeedsRebootReturnValue = installer.NewRebootActionPointer(testCase.moduleRebootReply)
```
Comment (kacf) [inline app/state_test.go:5449]: Can't you just do this?

```golang
c.FakeDevice.NeedsRebootReturnValue = &testCase.moduleRebootReply
```

File: app/state_test.go [hunk 5]
```diff
@@ -5405,3 +5409,53 @@ func TestFetchRetryUpdateControl(t *testing.T) {
...
+	tests := map[string]struct {
+		setInitialValue   bool
+		initialValue      datastore.RebootType
+		moduleRebootReply installer.RebootAction
+	}{
+		"Reboot already requested": {
+			setInitialValue:   true,
+			initialValue:      datastore.RebootTypeAutomatic,
+			moduleRebootReply: installer.RebootRequired,
+		},
```
Comment (kacf) [inline app/state_test.go:5430]: For completeness sake, maybe also test one where `moduleRebootReply: installer.NoReboot`? This is basically the old behavior, but doesn't hurt to verify that it still works.

**Review by kacf — APPROVED** [review]
Looks good, Only minor requests, so I'm pre-approving!

File: app/state_test.go (hunk 4 above)
Comment (oleorhagen) [reply app/state_test.go:5449]: _blush_

**Comment by oleorhagen** [comment]
@mender-test-bot start pipeline

**Comment by oleorhagen** [comment]
https://github.com/mendersoftware/integration/pull/1770

**Comment by oleorhagen** [comment]
@mender-test-bot start pipeline --pr integration/1770

**Comment by oleorhagen** [comment]
@mender-test-bot 
cherry-pick to:
* 3.1.x
* 3.0.x

**Pull request merged**